File: python/packages/kagent-adk/src/kagent/adk/_config_materialize.py
"""Materialize Agent Substrate configuration from environment variables.

The ActorTemplate injects config JSON directly and credentials through SecretKeyRef environment
variables. Credential placeholders are expanded before writing the files loaded by the ADK.
This mirrors the Go ADK's ``MaterializeFromEnv``.

When the environment variables are absent this is a no-op.
"""

import json
import logging
import os

logger = logging.getLogger(__name__)

# Environment variables injected by the substrate ActorTemplate, keyed to the file name the
# ADK loads from within the config directory.
_ENV_TO_CONFIG_FILE = {
    "KAGENT_CONFIG_JSON": "config.json",
    "KAGENT_AGENT_CARD_JSON": "agent-card.json",
}

# The bearer token is materialized to a fixed path outside the config dir, matching the Go ADK.
_KAGENT_TOKEN_ENV = "KAGENT_TOKEN"
_KAGENT_TOKEN_PATH = "/var/run/secrets/tokens/kagent-token"


def _materialize_env_to_file(env_key: str, path: str) -> bool:
    """Write the raw value of ``env_key`` to ``path`` (0600). Returns True if written."""
    value = os.getenv(env_key, "").strip()
    if not value:
        return False
    if env_key == "KAGENT_CONFIG_JSON" and "__KAGENT_ENV[" in value:
        value = json.dumps(_expand_config_env(json.loads(value)), separators=(",", ":"))
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        f.write(value)
    os.chmod(path, 0o600)
    return True


def _expand_config_env(value):
    if isinstance(value, str) and value.startswith("__KAGENT_ENV[") and value.endswith("]__"):
        name = value[len("__KAGENT_ENV[") : -len("]__")]
        if name not in os.environ:
            raise ValueError(f"required environment variable {name} is not set")
        return os.environ[name]
    if isinstance(value, list):
        return [_expand_config_env(item) for item in value]
    if isinstance(value, dict):
        return {key: _expand_config_env(item) for key, item in value.items()}
    return value


def materialize_from_env(config_dir: str) -> None:
    """Write substrate config environment variables to the paths the ADK loads from.

    No-op for any variable that is unset, so the volume-mounted Deployment path is unaffected.
    """
    for env_key, filename in _ENV_TO_CONFIG_FILE.items():
        if _materialize_env_to_file(env_key, os.path.join(config_dir, filename)):
            logger.info("Materialized %s from %s", filename, env_key)
    # Best-effort: the token path (/var/run/secrets/tokens) may not exist or be writable for a
    # nonroot runtime. A missing token only degrades authenticated callbacks, so log and continue
    # rather than crash startup.
    try:
        _materialize_env_to_file(_KAGENT_TOKEN_ENV, _KAGENT_TOKEN_PATH)
    except OSError as e:
        logger.warning("Could not materialize %s to %s: %s", _KAGENT_TOKEN_ENV, _KAGENT_TOKEN_PATH, e)

File: go/adk/pkg/config/config_materialize.go
package config

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const (
	envAgentConfigJSON = "KAGENT_CONFIG_JSON"
	envAgentCardJSON   = "KAGENT_AGENT_CARD_JSON"
	envSRTSettingsJSON = "KAGENT_SRT_SETTINGS_JSON"
	envKagentToken     = "KAGENT_TOKEN"
	kagentTokenDir     = "/var/run/secrets/tokens"
	kagentTokenFile    = "kagent-token"
	srtSettingsFile    = "srt-settings.json"
)

// MaterializeFromEnv writes Agent Substrate environment variables to
// the on-disk paths expected by the Go ADK runtime at startup.
func MaterializeFromEnv(configDir string) error {
	if err := materializeEnvToFile(envAgentConfigJSON, filepath.Join(configDir, "config.json")); err != nil {
		return fmt.Errorf("materialize agent config: %w", err)
	}
	if err := materializeEnvToFile(envAgentCardJSON, filepath.Join(configDir, "agent-card.json")); err != nil {
		return fmt.Errorf("materialize agent card: %w", err)
	}
	if err := materializeEnvToFile(envSRTSettingsJSON, filepath.Join(configDir, srtSettingsFile)); err != nil {
		return fmt.Errorf("materialize srt settings: %w", err)
	}
	if err := materializeEnvToFile(envKagentToken, filepath.Join(kagentTokenDir, kagentTokenFile)); err != nil {
		return fmt.Errorf("materialize kagent token: %w", err)
	}
	return nil
}

func materializeEnvToFile(envKey, path string) error {
	value := strings.TrimSpace(os.Getenv(envKey))
	if value == "" {
		return nil
	}
	if envKey == envAgentConfigJSON && strings.Contains(value, "__KAGENT_ENV[") {
		expanded, err := expandConfigEnv(value)
		if err != nil {
			return err
		}
		value = expanded
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return fmt.Errorf("create directory for %s: %w", path, err)
	}
	if err := os.WriteFile(path, []byte(value), 0o600); err != nil {
		return fmt.Errorf("write %s: %w", path, err)
	}
	if err := os.Chmod(path, 0o600); err != nil {
		return fmt.Errorf("chmod %s: %w", path, err)
	}
	return nil
}

func expandConfigEnv(raw string) (string, error) {
	var config any
	if err := json.Unmarshal([]byte(raw), &config); err != nil {
		return "", fmt.Errorf("decode config JSON: %w", err)
	}
	var expand func(any) (any, error)
	expand = func(value any) (any, error) {
		switch value := value.(type) {
		case string:
			const prefix, suffix = "__KAGENT_ENV[", "]__"
			if strings.HasPrefix(value, prefix) && strings.HasSuffix(value, suffix) {
				name := strings.TrimSuffix(strings.TrimPrefix(value, prefix), suffix)
				resolved, ok := os.LookupEnv(name)
				if !ok {
					return nil, fmt.Errorf("required environment variable %s is not set", name)
				}
				return resolved, nil
			}
		case []any:
			for i := range value {
				item, err := expand(value[i])
				if err != nil {
					return nil, err
				}
				value[i] = item
			}
		case map[string]any:
			for key, item := range value {
				expanded, err := expand(item)
				if err != nil {
					return nil, err
				}
				value[key] = expanded
			}
		}
		return value, nil
	}
	expanded, err := expand(config)
	if err != nil {
		return "", err
	}
	encoded, err := json.Marshal(expanded)
	if err != nil {
		return "", fmt.Errorf("encode config JSON: %w", err)
	}
	return string(encoded), nil
}
